gossipsub.go

```diff
@@ -1212,9 +1212,25 @@ func (gs *GossipSubRouter) Publish(msg *Message) {
 		if pid == from || pid == peer.ID(msg.GetFrom()) {
 			continue
 		}
-
+		// Add message to batch without sending it out
+		if msg.MessageBatch != nil {
+			msg.MessageBatch.AddRPC(pid, out)
+			continue
+		}
 		gs.sendRPC(pid, out, false)
 	}
+	if msg.MessageBatch != nil {
+		msg.MessageBatch.RemoveMessage(gs.p.idGen.ID(msg))
```

Comment:
Maybe we want to call `RemoveMessage` earlier. I think there's an issue where if your `gs.p.topics[topic]` is empty then `RemoveMessage` will never be called and the whole batch will never send. You can hit this if you are batching messages across topics and one topic happens to have no peers.